fms/models/gpt_bigcode.py

```diff
@@ -473,7 +474,94 @@ def _hf_sd_to_fms_sd(hf_sd: Mapping) -> Mapping:
     return new_sd
 
 
+def _gptq_fused_sd_to_fms_unfused_sd(hf_sd: Mapping) -> Mapping:
+    """Fused-checkpoint-to-unfused-GPTQ-model adapter.
+    Works for GPTQ models instantiated as unfused using unfused_strategy="pre"
+    """
+
+    replacements = [
+        ("lm_head.weight", "head.weight"),
+        (r"^transformer.wte.weight", "base_model.embedding.weight"),
+        (r"^transformer.wpe.weight", "base_model.position_embedding.weight"),
+        (r"^transformer.ln_f", "base_model.dec_norm"),
+        (r"^transformer.h", "base_model.layers"),
+        # need to do kqv manually
+        (r"attn\.c_proj", "attn.dense"),
+        (r"mlp\.c_fc", "ff_sub_layer.w1"),
+        (r"mlp\.c_proj", "ff_sub_layer.w2"),
+        (r"ln_1", "ln"),
+        (r"ln_2", "ff_ln"),
+    ]
+
+    pack_ratio = 8  # assume 4 bits => packing/compression ratio = 32 / 4
+
+    # derive out_feature from g_idx (instead of passing it to the adapter)
+    # TODO: improve this, it may be fragile...
+    for hf_name, param in hf_sd.items():
+        if "attn.c_attn.g_idx" in hf_name:
+            query_out_feat = param.size(0)
+            break
+
+    fms_unfused_sd = {}
+    for hf_name, param in hf_sd.items():
+        fms_name = hf_name
+        for pattern, repl in replacements:
+            fms_name = re.sub(pattern, repl, fms_name)
+
+        # extract qparams for unfused QKV model (prior any TP)
+        if "attn.c_attn" in hf_name:
+            query_fms_name = fms_name.replace("attn.c_attn", "attn.in_proj.query")
+            key_fms_name = fms_name.replace("attn.c_attn", "attn.in_proj.key")
+            value_fms_name = fms_name.replace("attn.c_attn", "attn.in_proj.value")
+
+            if "qweight" in fms_name:
+                in_feat_packed, out_feat = param.size()
+                kv_out_feat = (out_feat - query_out_feat) // 2
```

Comment:
I also assume that KV output features are identical. I believe this is always the case, even beyond gpt_bigcode models, isn't it?